Source/Particles/Deposition/ChargeDeposition.H

```diff
@@ -201,6 +201,323 @@ void doChargeDepositionShapeN (const GetParticlePosition& GetPosition,
+#endif
+
+            // Compute shape factor along x
+            // i: leftmost grid point that the particle touches
+            amrex::Real sx[depos_order + 1] = {0._rt};
+            int i = 0;
+            if (rho_type[0] == NODE) {
+                i = compute_shape_factor(sx, x);
+            } else if (rho_type[0] == CELL) {
+                i = compute_shape_factor(sx, x - 0.5_rt);
+            }
+#endif //defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ) || defined(WARPX_DIM_3D)
+#if defined(WARPX_DIM_3D)
+            // y direction
+            const amrex::Real y = (yp - ymin)*dyi;
+            amrex::Real sy[depos_order + 1] = {0._rt};
+            int j = 0;
+            if (rho_type[1] == NODE) {
+                j = compute_shape_factor(sy, y);
+            } else if (rho_type[1] == CELL) {
+                j = compute_shape_factor(sy, y - 0.5_rt);
+            }
+#endif
+            // z direction
+            const amrex::Real z = (zp - zmin)*dzi;
+            amrex::Real sz[depos_order + 1] = {0._rt};
+            int k = 0;
+            if (rho_type[WARPX_ZINDEX] == NODE) {
+                k = compute_shape_factor(sz, z);
+            } else if (rho_type[WARPX_ZINDEX] == CELL) {
+                k = compute_shape_factor(sz, z - 0.5_rt);
+            }
+
+            // Deposit charge into buf
+#if defined(WARPX_DIM_1D_Z)
+            for (int iz=0; iz<=depos_order; iz++){
+                amrex::Gpu::Atomic::AddNoRet(
+                    &buf(lo.x+k+iz, 0, 0, 0),
+                    sz[iz]*wq);
+            }
+#endif
+#if defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
+            for (int iz=0; iz<=depos_order; iz++){
+                for (int ix=0; ix<=depos_order; ix++){
+                    amrex::Gpu::Atomic::AddNoRet(
+                        &buf(lo.x+i+ix, lo.y+k+iz, 0, 0),
+                        sx[ix]*sz[iz]*wq);
+#if defined(WARPX_DIM_RZ)
+                    Complex xy = xy0; // Throughout the following loop, xy takes the value e^{i m theta}
+                    for (int imode=1 ; imode < n_rz_azimuthal_modes ; imode++) {
+                        // The factor 2 on the weighting comes from the normalization of the modes
+                        amrex::Gpu::Atomic::AddNoRet( &buf(lo.x+i+ix, lo.y+k+iz, 0, 2*imode-1), 2._rt*sx[ix]*sz[iz]*wq*xy.real());
+                        amrex::Gpu::Atomic::AddNoRet( &buf(lo.x+i+ix, lo.y+k+iz, 0, 2*imode  ), 2._rt*sx[ix]*sz[iz]*wq*xy.imag());
+                        xy = xy*xy0;
+                    }
+#endif
+                }
+            }
+#elif defined(WARPX_DIM_3D)
+            for (int iz=0; iz<=depos_order; iz++){
+                for (int iy=0; iy<=depos_order; iy++){
+                    for (int ix=0; ix<=depos_order; ix++){
+                        amrex::Gpu::Atomic::AddNoRet(
+                            &buf(lo.x+i+ix, lo.y+j+iy, lo.z+k+iz),
+                            sx[ix]*sy[iy]*sz[iz]*wq);
+                    }
+                }
+            }
+#endif
```

Comment:
Filter out additions of exact 0.0 because a particle exactly at a node/cell center on the upper bound may contribute with zero weight to a point beyond the bounds - this would let us save a guard cell